### PR TITLE
cephfs-mirror: bootstrap cephfs mirror daemon

### PR DESCRIPTION
**Pull request by vshankar**
This sets up the cephfs-mirror daemon to establish watch on a per-instance rados object and register itself (instance id) in cephfs-mirror object in omap. Most of the calls are asynchronous thereby classes are written in async-state machine fashion wherever possible.

Nothing super interesting right now -- daemon just bootstraps and acknowledges notifications sent via rados object. This can be tested sending a notify via rados cli:  `rados -p cephfs.a.meta notify cephfs_mirror.<instance-id> "ping"`. The daemon acks back an empty bufferlist.

### CODE REVIEW

**Comment by vshankar** [comment]
@mchangir I'll post an update this week for this PR, but let us start on reviews.

**Comment by vshankar** [comment]
minor note: I have some changes locally -- will push that out tomorrow. thx!

File: src/mds/FSMap.h [hunk 1]
```diff
@@ -36,6 +36,136 @@
...
+  bool is_mirrored() const {
+    return mirrored;
+  }
+  void enable_mirroring() {
+    mirrored = true;
+  }
+  void disable_mirroring() {
+    peers.clear();
+    mirrored = false;
+  }
```
Comment (mchangir) [inline src/mds/FSMap.h:118]: I'm not sure about the intended use of these methods, but `disable_mirroring()` clears the set of `peers` and sets `mirrored` to `false` and `enable_mirroring()` just sets `mirrored` back to `true`

Would it be a problem is the set of `peers` is empty when the mirroring is re-enabled ?

Comment (vshankar) [reply src/mds/FSMap.h:118]: err... you should probably review this commit in PR #36131

The idea is when these routines get called, fsmap epoch is increased and subscribers get notified with the new fsmap.

File: src/tools/cephfs_mirror/watcher/RewatchRequest.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#include "include/int_types.h"
+#include "include/rados/librados.hpp"
+
+struct Context;
+
+namespace cephfs {
+namespace mirror {
+namespace watcher {
+
+class RewatchRequest {
```
Comment (batrick) [inline src/tools/cephfs_mirror/watcher/RewatchRequest.h:17]: Please add a comment about what this is for.

**Comment by batrick** [comment]
(fantastic work Venky!!)

Also, please rebase :)

**Comment by vshankar** [comment]
> Organization looks really good. This looks nearly ready to merge. What I'd like to see though is a brief description of how the classes are organized to help newcomers with understanding the layout.

ack -- will push an update today. thx!

**Comment by vshankar** [comment]
rebased and updated. please take a look. cephfs-mirror is build by default, i.e. w/ WITH_CEPHFS config.

**Comment by sebastian-philipp** [comment]
You also want to add a cephadm integration?

**Comment by batrick** [comment]
> You also want to add a cephadm integration?

Yes. @vshankar please make an issue for that so we don't forget.

**Comment by dmick** [comment]
Looks like this adds a binary cephfs-mirror, but does not update packaging files, so the master build is broken (centos8/leap validate that built binaries are packaged).

**Comment by tchaikov** [comment]
@batrick and @vshankar how did you test this change? as @dmick pointed out. it broke the rpm build.

**Comment by batrick** [comment]
Reverted in https://github.com/ceph/ceph/pull/36934

**Comment by batrick** [comment]
@tchaikov sorry. There are no tests yet for this so I thought a `make check` success was sufficient. That skipped packaging done by ceph-ci which is what failed. Lesson learned for me.

**Comment by vshankar** [comment]
> You also want to add a cephadm integration?

https://tracker.ceph.com/issues/47261

@sebastian-philipp I have assigned this to you for now...

**Comment by sebastian-philipp** [comment]
> > You also want to add a cephadm integration?
> 
> https://tracker.ceph.com/issues/47261
> 
> @sebastian-philipp I have assigned this to you for now...

Ha. Very well then. Note that this is rather low on my priority list. If you want to have this in a reasonable time frame, you might want to ask someone else. I could give you an in-depth guide of of how to add a new service type.

**Comment by vshankar** [comment]
> Ha. Very well then. Note that this is rather low on my priority list. If you want to have this in a reasonable time frame, you might want to ask someone else. I could give you an in-depth guide of of how to add a new service type.

ACK. Appreciate any docs for the same.

**Pull request merged**